Convert TimePicker to Material 3

gspencergoog opened the pull request:
## Description

This converts the time picker to use the M3 spec.

<img width="912" alt="Screen Shot 2022-12-01 at 7 35 22 PM" src="https://user-images.githubusercontent.com/8867023/205209274-65a104aa-66f3-45de-8556-67f2ea7ea56e.png">

In the process, I had to rearrange a lot of the architecture of the `TimePicker`, since it didn't lend itself to the kinds of layout that the M3 spec wanted.

This PR includes these changes:
* Add a `_TimePickerDefaultsM2` class that contains the defaults for the M2 version of the picker using M3 attributes.
* Add a base `_TimePickerDefaults` class that both `_TimePickerDefaultsM2` and `_TimePickerDefaultsM3` derive from so that they can always return non-nullable values.
* Instead of threading a number of modifiable parameters through all of the widgets, use a single `InheritedModel` that contains the state for the picker.
* Add the M3 defaults class `_TimePickerDefaultsM3` and the appropriate generation template.
* Fixed layout issues where small window sizes caused overflow exceptions.  It resizes nicely now until it reaches a minimum size, and the scrolls the contents after that.
* Added a bunch of material3 specific tests.
* Added a `show_time_picker.0.dart` usage example that allows configuring of the time picker before showing it.
* Reorganized things so that there's a `_TimePicker` widget that could be made public if we want to allow embedding of time pickers into other widgets (instead of in a dialog).

I had to "invent" the token values for the text input fields of the "time input" controls, since while there are values for the hour/minute selector, there's nothing in there for the input field focus colors, borders, etc.

The updated picker handles small windows much better...
<video src="https://user-images.githubusercontent.com/8867023/205209754-8fb43ff5-a65a-4493-ac01-fda1d087ac2a.mp4" controls="controls" style="max-width: 730px;">
</video>

## Related Issues
 - Fixes https://github.com/flutter/flutter/issues/101480

## Tests
 - Added tests for many Material3 attributes and layouts.
 - Added a test for the new example.

gspencergoog commented:
@darrenaustin Feel free to start reviewing this: it's only a draft because it's missing a test for the example.

I'm sure you'll find plenty of other things to review here... Sorry for the size.

gspencergoog commented:
Actually, I think I'll try and factor out and land some of the unrelated changes before I take this off of "draft".

gspencergoog committed:
Make some minor changes in preparation for updating the Time Picker to M3

gspencergoog committed:
Convert TimePicker to Material 3

gspencergoog committed:
Add example test

gspencergoog committed:
Revert OutlineInputBorder.borderRadius type change

gspencergoog committed:
Revert OutlineInputBorder.borderRadius type change

gspencergoog committed:
Revert more OutlineInputBorder.borderRadius changes.

gspencergoog commented:
Okay, I've factored out and landed the unrelated changes.

darrenaustin commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -2534,3 +3049,664 @@ Future<TimeOfDay?> showTimePicker({
   SemanticsService.announce(message, Directionality.of(context));
 }
+
+// An abstract base class for the M2 and M3 defaults below, so that their return
+// types can be non-nullable.
+abstract class _TimePickerDefaults extends TimePickerThemeData {
```
Interesting. In the other components we just went with a direct subclass of the component theme data. For the few cases where the default had to end in a non-null value, we would just use the '!' operator when pulling it from the defaults. Not as clean perhaps, but reduces the amount of bloat in having yet another full subclass definition in each component.

However, in this case it looks like you are also adding new fields that aren't part of the theme data. Is there some reason those new fields shouldn't be part of the public theme?

gspencergoog replied:
I made the separate class with non-null fields so that I could use it everywhere as the type of the resolved default theme, then I wouldn't need to use the `!` everwhere that the default theme is used. It felt a lot cleaner to me, and the intermediate class is pretty much empty.

The sizes seemed like either a bunch of sizes that weren't part of the spec as being customizable, but were part of the token information, or which consolidated the sizing information that was encoded as a bunch of constants in the file that had to be different for M2 and M3, so it seemed better to just use the defaults for those, since they have to be different for M2 and M3 as well.

So, instead of:

```dart
const double _kDayPeriodInputWidth = xx;
const double _kDayPeriodInputHeight = xx;
const double _kDayPeriodInputWidthM2 = xx;
const double _kDayPeriodInputHeightM2 = xx;
```

I could just refer to the sizes in the defaults.

darrenaustin replied:
That makes sense. Early on in the migration we got some push back on adding more classes to each widget as it was inflating the download size for the web. I believe each dart class gets transpiled into a JS class, so that could be significant for these boilerplate heavy classes. That said in this case it probably doesn't matter that much and ti is cleaner. Thx.

darrenaustin commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -1915,168 +2178,491 @@ class TimePickerDialog extends StatefulWidget {
+      case TimePickerEntryMode.inputOnly:
+        final _TimePickerDefaults defaultTheme;
+        final bool alwaysUse24HourFormat = MediaQuery.of(context).alwaysUse24HourFormat;
+        if (theme.useMaterial3) {
+          defaultTheme = _TimePickerDefaultsM3(context);
+          double timePickerWidth;
+          if (alwaysUse24HourFormat) {
+            timePickerWidth = _kTimePickerInputSize.width - defaultTheme.dayPeriodPortraitSize.width - 12;
+          } else {
+            timePickerWidth = _kTimePickerInputSize.width;
+          }
+          timePickerSize = Size(timePickerWidth, _kTimePickerInputSize.height);
+        } else {
+          defaultTheme = _TimePickerDefaultsM2(context);
+          double timePickerWidth;
+          if (alwaysUse24HourFormat) {
+            timePickerWidth = _kTimePickerInputSize.width - defaultTheme.dayPeriodPortraitSize.width - 12;
+          } else {
+            timePickerWidth = _kTimePickerInputSize.width;
+          }
+          timePickerSize = Size(timePickerWidth, _kTimePickerInputSize.height);
+        }
```
This looks duplicated. The only difference between M2 and M3 is just the defaults, right? 

Can't we just use the right default theme (M2/M3) and use the same block of code for both M3 and M2? Or am I missing something that is different in the two cases?

gspencergoog replied:
Good catch, I didn't notice that!  I consolidated it.

darrenaustin commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -1915,168 +2178,491 @@ class TimePickerDialog extends StatefulWidget {
+  }
+}
+
+// The _TimePicker widget is constructed so that in the future we could expose
+// this as a public API for embedding time pickers into other non-dialog
+// widgets, once we're sure we want to support that.
```
Nice. We have in the past seen similar requests for exposing parts of the date picker as well.

gspencergoog replied:
Yeah, I figured I would arrange it this way, and if/when we decide to make it available seprately, we could do the date picker at the same time.

darrenaustin commented on packages/flutter/lib/src/material/time_picker_theme.dart:
```diff
@@ -133,117 +162,148 @@ class TimePickerThemeData with Diagnosticable {
-  /// If this is null, the time picker defaults to the overall theme's
-  /// [TextTheme.titleMedium].
-  final TextStyle? dayPeriodTextStyle;
+  /// Deprecated. Use [headlineTextStyle] instead.
+  @Deprecated(
+    'Use headlineTextStyle instead. '
+    'This feature was deprecated after v3.6.0-0.0.pre.',
+  )
+  final TextStyle? helpTextStyle;
```
Do we have a flutter fix for this and a migration guide? Or do we want to handle that in a separate PR? Or is this an existing deprecation that just got moved around in this cleanup? 😀

gspencergoog replied:
No, it's a new deprecation, and I think I'd rather handle the flutter fix in another PR.
Or, I could just duplicate the value here, and deprecate it in the PR that has the fix.

gspencergoog replied:
Actually, I'll just not deprecate it at all in this PR.  We can do that later.

darrenaustin replied:
Either would be fine. I just wanted to make sure we have a plan to follow through on the deprecation. Thx.

darrenaustin commented on packages/flutter/test/material/time_picker_test.dart:
```diff
@@ -358,989 +234,1240 @@ void _tests() {
 
-    semantics.dispose();
-  });
+    testWidgets('Material2 has ALL CAPS labels', (WidgetTester tester) async {
```
For M2 specific tests it would be good to hardwire `useMaterial3: false` so that when we change the default to true they will still work properly.

gspencergoog replied:
Oh, good point. Done.

darrenaustin commented on packages/flutter/test/material/time_picker_test.dart:
```diff
@@ -358,989 +234,1240 @@ void _tests() {
+      });
+    });
+
+    group('Works for various view sizes', () {
```
Nice!

darrenaustin reviewed (commented):
Wow, nice work! There is a lot here, and some of the diffs were a little hard to follow, so my apologies if I missed something. Some questions and comments below. Thx!

gspencergoog committed:
Fix test

gspencergoog commented on dev/tools/gen_defaults/lib/time_picker_template.dart:
```diff
@@ -0,0 +1,363 @@
+''';
+
+  /// Generate a [BorderSide] for the given components.
+  String mergedBorder(String componentToken1, String componentToken2) {
```
Hmm, I guess not.  Removed.

gspencergoog commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -86,186 +79,203 @@ enum TimePickerEntryMode {
-                    const SizedBox(width: 12.0),
-                  ],
-                  Expanded(
-                    child: Row(
-                      // Hour/minutes should not change positions in RTL locales.
-                      textDirection: TextDirection.ltr,
-                      children: <Widget>[
-                        Expanded(child: _HourControl(fragmentContext: fragmentContext)),
-                        _StringFragment(timeOfDayFormat: timeOfDayFormat),
-                        Expanded(child: _MinuteControl(fragmentContext: fragmentContext)),
-                      ],
-                    ),
+            Row(
+              children: <Widget>[
+                if (hourDialType == _HourDialType.twelveHour &&
+                    timeOfDayFormat == TimeOfDayFormat.a_space_h_colon_mm) ...<Widget>[
+                  const _DayPeriodControl(),
+                ],
+                Expanded(
+                  child: Row(
+                    // Hour/minutes should not change positions in RTL locales.
+                    textDirection: TextDirection.ltr,
+                    children: <Widget>[
+                      const Expanded(child: _HourControl()),
+                      _StringFragment(timeOfDayFormat: timeOfDayFormat),
+                      const Expanded(child: _MinuteControl()),
+                    ],
                   ),
-                  if (!use24HourDials && timeOfDayFormat != TimeOfDayFormat.a_space_h_colon_mm) ...<Widget>[
-                    const SizedBox(width: 12.0),
-                    _DayPeriodControl(
-                      selectedTime: selectedTime,
-                      orientation: orientation,
-                      onChanged: onChanged,
-                    ),
-                  ],
+                ),
+                if (hourDialType == _HourDialType.twelveHour &&
+                    timeOfDayFormat != TimeOfDayFormat.a_space_h_colon_mm) ...<Widget>[
+                  const SizedBox(width: 12),
```
Reformatted.

gspencergoog commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -86,186 +79,203 @@ enum TimePickerEntryMode {
-      onModeChanged(value);
+  @override
+  bool updateShouldNotifyDependent(_TimePickerModel oldWidget, Set<_TimePickerAspect> dependencies) {
+    if (use24HourFormat != oldWidget.use24HourFormat && dependencies.contains(_TimePickerAspect.use24HourFormat)) {
```
Oh, absolutely.

gspencergoog committed:
Review Changes

gspencergoog commented:
Thanks for that huge review!

Okay, I think I addressed all the comments, PTAL. For the test, I actually rearranged it so that a) it's explicit whether things are set up as material2 or material3, and I put a loop around the test definitions so that we actually run every test once in material2 mode and once in material3 mode.

darrenaustin reviewed (approved):
![flutter-lgtm](https://user-images.githubusercontent.com/19588/202313172-687ccda4-96b5-47c4-b8e5-fe653dd442c7.png)

Nice work. Thx.

gspencergoog committed:
Merge changes

gspencergoog committed:
Some sizing and elevation fixes

gspencergoog committed:
Fix localization tests
